### DISCUSSION TITLE
Question: Best Practice for Storing Temporary Keys for Escrow Contract

### DISCUSSION THREAD

**Discussion by jrweinb** [celo-org/social-connect#221] [Help (Q&A)] [answered]
Hi - I am considering Celo for a project and the escrow contract is very appealing. I am wondering the best practice around storing temporary private keys. In a database? Any help is much appreciated!

Thanks.

**Comment by arthurgousset** [comment]
Adding a [link](https://discord.com/channels/600834479145353243/600840423958904842/946044628409397288) to this question on Discord (for future reference).

**Comment by critesjosh** [comment]
It depends on how you are using the escrow contract and sharing keys. How are you sharing the key with users? How much value will be secured by these keys? 

It might be good to store the keys encrypted in a database so if they database gets compromised then the keys are safe. You'll need to be very careful with the decryption key(s), maybe use an [HSM](https://en.wikipedia.org/wiki/Hardware_security_module).

**Comment by arthurgousset** [comment]
I pinged @isabellewei and she will get back to you asap. She will be best-placed to answer your question :)

**Comment by isabellewei** [comment, marked as answer]
Hi @jrweinb! Apologies that this isn't explained in the docs, but the new escrow flow in Valora no longer requires you to generate a temporary private key for the `paymentId` and send it to the receiver. Instead, the phone hash and pepper is used to generate the `paymentId`. Instead of sending the `paymentId` to the invited user, when claiming escrowed transactions Valora uses their phone number to compute the paymentId, so the invited user just needs to download Valora and verify their number

you can see the function that generates the paymentId: https://github.com/celo-org/wallet/blob/623bf8c483fb2fdcde6e6425070c021dfd8d3834/packages/mobile/src/escrow/utils.ts#L22-L37 and where it is used: https://github.com/celo-org/wallet/blob/623bf8c483fb2fdcde6e6425070c021dfd8d3834/packages/mobile/src/escrow/saga.ts#L154

**Comment by arthurgousset** [comment]
Quick ping because I transferred this discussion 

- from: `celo-org/celo-monorepo/discussions`
- to: `celo-org/identity/discussions`

to be more flexible in the types of discussions we have:

- 💡 [Feature requests (Feedback)](https://github.com/celo-org/identity/discussions/categories/feature-requests-feedback)
- 😺 [Help needed (Q&A)](https://github.com/celo-org/identity/discussions/categories/help-needed-q-a)
- 🗳️ [Polls (Voting)](https://github.com/celo-org/identity/discussions/categories/polls-voting)
- 🙌 [Show and tell (Updates)](https://github.com/celo-org/identity/discussions/categories/show-and-tell-updates)
 
Sorry for any inconvenience 🙏
